software/contrib/quantizer.py

```diff
@@ -0,0 +1,537 @@
+            """
+            if self.mode == MODE_TRIGGERED:
+                cv6.off()
+            
+        @b1.handler
+        def on_b1_press():
+            """Handler for pressing button 1
+
+            Button 1 is used for the main interaction and is passed to
+            the current display for user interaction
+            """
+            self.last_interaction_time = time.ticks_ms()
+            self.active_screen.on_button1()
+            
+        @b2.handler
+        def on_b2_press():
+            """Handler for pressing button 2
+
+            Button 2 is used to cycle between screens
+            """
+            self.last_interaction_time = time.ticks_ms()
+            
+            if self.active_screen == self.kb:
+                self.active_screen = self.menu
+            else:
+                self.active_screen = self.kb
+        
+    def load(self):
+        """Load the persistent settings from storage and apply them
+        """
+        state = self.load_state_json()
+        
+        loaded_scale = state.get("scale", [True]*12)  # default to a chromatic scale
+        self.scale.notes = loaded_scale
+        
+        self.root = state.get("root", self.root)
+        self.octave = state.get("octave", self.octave)
+        self.intervals = state.get("intervals", self.intervals)
+        self.mode = state.get("mode", self.mode)
+    
+    def save(self):
+        """Save the current settings to persistent storage
+        """
+        state = {
+            "scale": self.scale.notes,
+            "root": self.root,
+            "octave": self.octave,
+            "intervals": self.intervals,
+            "mode": self.mode
+        }
+        self.save_state_json(state)
+        
+    @classmethod
+    def display_name(cls):
+        return "Quantizer"
+       
+    def quantize(self, analog_in):
+        """Take an analog signal and process it
+
+        Sets self.current_note and self.output_voltage
+        """
+        
+        (volts, semitone) = self.scale.quantize(analog_in)
+        
+        # apply our octave & transposition offsets
+        volts = volts + self.octave * VOLTS_PER_OCTAVE + self.root * VOLTS_PER_SEMITONE
+        
+        self.output_voltage = volts
+        self.current_note = semitone
+       
+    def read_quantize_output(self):
+        """Read the input signal, quantize it, set outputs 1-5 accordingly
+
+        Called by the main loop in continuous mode or the rising clock handler
+        in triggered mode
+        """
+        self.input_voltage = ain.read_voltage(128)
+        self.quantize(self.input_voltage)
+        
+        cv1.voltage(self.output_voltage)
+        
+        for i in range(len(self.aux_outs)):
+            self.aux_outs[i].voltage(self.output_voltage + self.intervals[i] * VOLTS_PER_SEMITONE)
+    
+    def main(self):
+        """The main loop; reads from ain, sets the output voltages
+        """
+        
+        while True:
+            # Update at 100Hz
```

Comment:
[discussion] I feel like this comment is misleading. I don't think that we know enough about the rate of execution of the rest of the loop to make a claim about how fast we are quantizing.